pkg/logql/syntax/parser.go

```diff
@@ -134,6 +134,29 @@ func ParseMatchers(input string) ([]*labels.Matcher, error) {
 	return matcherExpr.Mts, nil
 }
 
+// ExtractMatchersFromQuery extracts the matchers from a query.
+// compared to ParseMatchers, it will not fail if the query contains anything else.
+func ExtractMatchersFromQuery(input string) ([]*labels.Matcher, error) {
```

Comment:
Good call! I didn't think about the queries that may contain multiple matchers. E.g.
```
count_over_time({job="foo"}[5m]) / count_over_time({job="bar"}[5m] offset 10m)
```

IIUC, similarly to in `shard_resolver.go`, we need to create a IndexStats request for each matcher. Then we need to sum the bytes for each IndexStats response. That's the amount of data a query would read.

Reply:
I implemented this in [2a1df3e](https://github.com/grafana/loki/pull/8670/commits/2a1df3ee0913e97f031ffa64419552d5d841a3fe). Looks like it works fine. If this is what you meant, I do have some new questions 😅:
- Shall we set the concurrency to get index stats from a config as done in [shardResolverForConf][2]?
  - I can see some places where the `ForEach` is set to [nonconfigurable values][3].
  - For simplicity, I think it should be ok to leave the concurrency as `len(matcherGroups)` or to a fixed not-so-high value (e.g. 10). But happy to change my mind.
- As in [dynamicShardResolver.Shards][4], for log selector queries, should we also subtract the `max_look_back_period` from the query start time? I think I don't quite understand why is that needed on sharding.


[Tested in loki-dev-009][1]

[4]: https://github.com/grafana/loki/blob/6a7403c4f5cd0a01dea13954fe6f51631f485ad1/pkg/querier/queryrange/shard_resolver.go#L87-L89
[3]: https://github.com/grafana/loki/blob/6bf782b0a784e0aeeb36c9e292532d631955f134/pkg/ruler/base/ruler.go#L893
[2]: https://github.com/grafana/loki/blob/6a7403c4f5cd0a01dea13954fe6f51631f485ad1/pkg/querier/queryrange/shard_resolver.go#L42
[1]: https://admin-dev-eu-west-2.grafana.net/grafana/explore?orgId=1&left=%7B%22datasource%22:%22loki-dev-009-local%22,%22queries%22:%5B%7B%22refId%22:%22A%22,%22expr%22:%22sum%28%5Cn%20%20%20%20sum_over_time%20%28%5Cn%20%20%20%20%20%20%20%20%7Bcluster%3D%5C%22dev-eu-west-2%5C%22,%20namespace%3D%5C%22loki-dev-009%5C%22,%20container%3D%5C%22distributor%5C%22%7D%5Cn%20%20%20%20%20%20%20%20%7C%3D%20%5C%22totalSize%3D%5C%22%5Cn%20%20%20%20%20%20%20%20%7C%20pattern%20%5C%22%3C_%3E%20totalSize%3D%5C%5C%5C%22%3CtotalSize%3E%5C%5C%5C%22%20%3C_%3E%5C%22%5Cn%20%20%20%20%20%20%20%20%7C%20unwrap%20bytes%28totalSize%29%5Cn%20%20%20%20%20%20%20%20%5B1h%5D%5Cn%20%20%20%20%29%5Cn%20%20%20%20%2F%5Cn%20%20%20%20sum_over_time%20%28%5Cn%20%20%20%20%20%20%20%20%7Bcluster%3D%5C%22dev-eu-west-2%5C%22,%20namespace%3D%5C%22loki-dev-009%5C%22,%20container%3D%5C%22distributor%5C%22%7D%5Cn%20%20%20%20%20%20%20%20%7C%3D%20%5C%22streamLabelsSize%3D%5C%22%5Cn%20%20%20%20%20%20%20%20%7C%20pattern%20%5C%22%3C_%3E%20streamLabelsSize%3D%5C%5C%5C%22%3CstreamLabelsSize%3E%5C%5C%5C%22%20%3C_%3E%5C%22%5Cn%20%20%20%20%20%20%20%20%7C%20unwrap%20bytes%28streamLabelsSize%29%5Cn%20%20%20%20%20%20%20%20%5B2h%5D%20offset%2010m%5Cn%20%20%20%20%29%5Cn%29%22,%22queryType%22:%22range%22,%22datasource%22:%7B%22type%22:%22loki%22,%22uid%22:%22loki-dev-009-local%22%7D,%22editorMode%22:%22code%22,%22hide%22:false%7D%5D,%22range%22:%7B%22from%22:%22now-15d%22,%22to%22:%22now%22%7D%7D

Reply:
`max_look_back_period` is generally most important in limited instant queries (which are rarely used). If I submit `{job="foo"}` with `start=end`, It would always return zero results. Prometheus has a concept of "look back <x> amount of time for instant queries" since metric data is sampled at some configurable `scrape_interval` (commonly 15s, 30s, or 1m). We sort of copy that idea and say, "find me logs from the past <lookback> when start=end".

Reply:
Just came across this myself. I don't think we need an extra method.